### DISCUSSION TITLE
Content in embedded files (via Snippets) doesn't show up in search result previews

### DISCUSSION THREAD

**Discussion by miguno** [squidfunk/mkdocs-material#2313] [Q&A]
**Background**
I have a Markdown file that embeds a local HTML file, using [Snippets](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/) (see [Material docs on embedding files](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#embedding-external-files)). 

```markdown
# Configuration

## Embedding HTML files

!!! success "Embedding external files is supported"
    Everything **below** this line is the embedded output of an embedded HTML file.

--8<-- "documentation/external.html"
```

Searching this page works. However, search result _previews_ only work for content that is stored directly in the Markdown file. Example of successfully searching for "embedding":

<img width="756" alt="search_preview_works" src="https://user-images.githubusercontent.com/294849/108495965-bd8b6200-72a9-11eb-9f1d-d59938e2cd3b.png">

Searching for terms in the embedded HTML file partially works, as the parent page ("Configuration") shows up, but without the useful search result preview as in the previous screenshot. Example of searching for "hostname", which is from the text of the embedded HTML file:

<img width="755" alt="search_preview_fails" src="https://user-images.githubusercontent.com/294849/108496394-41454e80-72aa-11eb-9472-12595bd5a015.png">

Some of my attempts to fix this issue included:

- Enabling `prebuild_index`, followed by `mkdocs build` and serving the output via Apache httpd (nope).
- Adding the external HTML file to `nav:` in `mkdocs.yml` (nope). I thought, apparently mistakenly, that the search indexing might be limited to what's explicitly listed there.

**Environment**

* mkdocs v1.1.2
* mkdocs-material v6.2.8
* mkdocs-material-extensions v1.0.1

Tested with both `mkdocs serve` (built-in webserver) and serving the output of `mkdocs build` via Apache httpd.

**My question:** How can I get the search result previews to work also for embedded files?

**Comment by squidfunk** [comment]
@facelessuser any idea? I thought Snippets was run as preprocessor, so MkDocs shouldn't see the Snippets but the replaced text?

**Reply by facelessuser** [reply]
It is run early in the Markdown parsing. Close to one of the first things. I assume search index is built on the HTML after Markdown parsing.

I suspect we are missing some info here. I doubt Snippets is responsible for whatever issue is being experienced.

I literally just woke up, so you'll have to give me a bit, but I'll provide some examples that demonstrate Snippets work in search.

**Reply by facelessuser** [reply]
Consider this Snippet: https://github.com/facelessuser/wcmatch/blob/master/docs/src/markdown/_snippets/posix.txt. I've created a Snippet that contains a table for POSIX character classes that is referenced on multiple pages: `glob` and `fnmatch`.

Here is an image of search working in that snippet:

<img width="711" alt="Screen Shot 2021-02-19 at 7 01 17 AM" src="https://user-images.githubusercontent.com/1055125/108513847-5ad1a080-7280-11eb-9bb1-e0b0081c4ef5.png">

You can also try it out live: https://facelessuser.github.io/wcmatch/.

So, search indexes work in snipppets. If search terms are not showing up that are included in a snippet, I would suspect there is something else at work here outside of it simply being a snippet.

**Reply by miguno** [reply]
Thanks for the quick feedback, @squidfunk and @facelessuser. I looked at @facelessuser's setup linked above, where it does indeed work as expected.

> I would suspect there is something else at work here outside of it simply being a snippet.

It turns out this is true. My experiments have thus far surfaced the following: HTML headings tags like `<h1>`, `<h2>` cause the problem.

The following simple snippet files (3 lines each) all result in successfully working search result previews when searching for "jupiter", "saturn", or "nepture":

```html
<code>jupiter</code>
saturn
neptune
```

```html
<div>jupiter</div>
saturn
neptune
```
```html
<p>jupiter</p>
saturn
neptune
```

```html
<head>jupiter</head>
saturn
neptune
```

However, any inclusion of `<h1>` et al. causes the problem of missing search result previews when searching for "jupiter", "saturn", or "neptune":

```html
<h1>jupiter</h1>
saturn
neptune
```

PS: I used code blocks in my reply here, but apparently the rendering (styling) doesn't work correctly in a response comment.

**Reply by facelessuser** [reply]
So, in general, only one `h1` should be in your document. You can have as many `h2`+ as you'd like, but `h1` should be restricted to one and is used as your title often.

**Reply by squidfunk** [reply]
Note that search will first try to highlight a match in a subsection (`h2`, `h3`, `h4`...), and if it doesn't find it, use the `h1` as a fallback. Furthermore, the search result summaries are not optimal as they only show the beginning of the text, so if the match is in a later portion, it won't get highlighted, i.e. there are not "..." in the summaries for non-matching parts of the text.

A match in the text directly after the `h1` and _before_ any further `hx`:

<img width="713" alt="Bildschirmfoto 2021-02-19 um 16 29 05" src="https://user-images.githubusercontent.com/932156/108524739-a8c1c500-72cf-11eb-8e3b-4588b39223a3.png">

A match in any of the subsections:

<img width="705" alt="Bildschirmfoto 2021-02-19 um 16 29 17" src="https://user-images.githubusercontent.com/932156/108524830-c2630c80-72cf-11eb-8a9f-71e16ef93a17.png">

Improving search result summaries is on the [roadmap](https://squidfunk.github.io/mkdocs-material/insiders/#future).

@facelessuser thanks for investigating.

**Reply by miguno** [reply]
> So, in general, only one h1 should be in your document. You can have as many h2+ as you'd like, but h1 should be restricted to one and is used as your title often.

To clarify: This was for testing. I tested with h1, h2, h3, h4, h5, etc. — all have the same result that previews are broken. And it's broken not just for the text between the `<hX>` tags, but everything else that's in the snippet file.

**Reply by squidfunk** [reply]
Okay. Then, please file an issue with a minimal reproducible example, starting from a mint Material for MkDocs installation (i.e. minimal `mkdocs.yml` and no overrides). If we have an example to reproduce the behavior and know exactly what you're aiming for we might either fix the issue or help you make it work. Also please explain exactly what you mean with "broken".

**Reply by miguno** [reply]
@squidfunk:
> Note that search will first try to highlight a match in a subsection (h2, h3, h4...), and if it doesn't find it, use the h1 as a fallback. Furthermore, the search result summaries are not optimal as they only show the beginning of the text, so if the match is in a later portion, it won't get highlighted, i.e. there are not "..." in the summaries for non-matching parts of the text.

Ack, thanks for the info.

But that would still mean that, in my example, "neptune" and "saturn" (plus even "jupiter") would still be expected to show up, no?

**Reply by facelessuser** [reply]
Yeah, I have no trouble under `h2`, `h3`, etc., but I assume there is yet something else that triggers this case. I believe I have seen the described behavior before in some cases, but I cannot seem to put my finger on the exact scenario that caused it.

**Comment by miguno** [comment]
@facelessuser : Here is a minimal example setup that reproduces the issue. The reproduction isn't 100% to what I described above, but it's close enough and seems to show the underlying cause. README includes description of what "broken" means.

https://github.com/miguno/mkdocs-search-previews

**Reply by miguno** [reply]
I will file an issue separately.

**Reply by facelessuser** [reply]
Your header doesn't have an id, so I imagine that is why it doesn't link it.

**Reply by squidfunk** [reply]
Could you please check whether the issue persists when literal `hx` tags are inserted directly into the page, without using Snippets?

**Reply by facelessuser** [reply]
Oh, Jupiter also doesn't make it to the Toc. I think raw HTML headers aren't getting handled by Python Markdown and MkDocs the same.

**Reply by facelessuser** [reply]
Yup, changing to `### Jupiter` works.

**Reply by facelessuser** [reply]
It appears that TOC doesn't handle raw HTML headers.

**Reply by squidfunk** [reply]
If it's related to and can be isolated to TOC, we must push this upstream to Python Markdown.

**Comment by miguno** [comment]
Issue created at https://github.com/squidfunk/mkdocs-material/issues/2315
